ang/civicase/case/details/people-tab/directives/case-details-people-tab.directive.js

```diff
@@ -160,59 +185,77 @@
     };
 
     /**
-     * Assign the role to a contact
+     * Prompts the user to create either a new role or client related to the case.
      *
-     * @param {string} role role
-     * @param {boolean} replace replace
+     * @param {Role} role the role to assign to the case.
      */
-    $scope.assignRole = function (role, replace) {
-      var message = '<input name="caseRoleSelector" placeholder="' + ts('Select Contact') + '" />';
-      if (role.role !== 'Client') {
-        message = message + '<br/><textarea rows="3" cols="35" name="description" class="crm-form-textarea" style="margin-top: 10px;padding-left: 10px;border-color: #C2CFDE;color: #9494A4;" placeholder="Description"></textarea>';
+    $scope.assignRoleOrClient = function (role) {
+      var isAssigningRole = role && !!role.relationship_type_id;
+
+      if (isAssigningRole) {
+        assignRole(role);
+      } else {
+        assignClient();
       }
-      CRM.confirm({
-        title: replace ? ts('Replace %1', { 1: role.role }) : ts('Add %1', { 1: role.role }),
-        message: message,
-        open: function () {
-          $('[name=caseRoleSelector]', this).crmEntityRef({ create: true, api: { params: { contact_type: role.contact_type, contact_sub_type: role.contact_sub_type }, extra: ['display_name'] } });
-        }
-      }).on('crmConfirm:yes', function () {
-        var params;
-        var apiCalls = [];
-        var val = $('[name=caseRoleSelector]', this).val();
-        var desc = $('[name=description]', this).val();
-        if (replace) {
-          apiCalls.push(unassignRoleCall(role));
-        }
-        if (val) {
-          var newContact = $('[name=caseRoleSelector]', this).select2('data').extra.display_name;
-          // Add case role
-          if (role.relationship_type_id) {
-            params = {
-              relationship_type_id: role.relationship_type_id,
-              start_date: 'now',
-              end_date: null,
-              contact_id_b: val,
+    };
+
+    /**
+     * Replaces the given role or client with another one selected by the user.
+     *
+     * @param {Role} role the role to replace.
+     */
+    $scope.replaceRoleOrClient = function (role) {
```

Comment:
This function has become too large to follow. I suggest the following refactoring,
1. Lets move the `promptForContact` callback to a separate function.
2. `activitySubject` and `apiCalls` can also be moved to separate functions.
```
activitySubject = getActivitySubject();
apiCalls = getApiCalls();
```

Reply:
Moved the apiCalls to different functions for replacing the case client, or a case role.